File: Crawler/controller.py
from frontier import Frontier, FrontierRevisit
from worker import CrawlerThread, RevisiterThread
from storage import Storage
from dashboard import Dashboard
from periodic import PeriodicThread


class Controller:
    """The Controller of the hole Crawling process"""

    def __init__(self, num_threads, seeds, cont_to_crawl):
        self.num_workers = num_threads
        self.dash = Dashboard(num_threads)
        self.workers = []
        self.frontier = Frontier(num_threads, self.dash)
        self.db = Storage()

        # Create the workers
        for i in range(num_threads):
            self.workers.append(CrawlerThread(i, 'CrawlerThread' + str(i), self.frontier, self.dash))
        # print("Workers created")
        # insert seeds in to serve
        if not cont_to_crawl:
            self.frontier.push_to_serve(seeds, 0)
            # print("seeds pushed")
            self.frontier.distribute()
            # print("seeds distributed")
        else:
            self.frontier.load_to_crawl()

    def run(self):
        """The main Program"""
        try:
            for i in range(self.num_workers):
                self.workers[i].start()
            # print("All Workers started")
            self.saver_to_crawl = PeriodicThread(self.frontier.save_to_crawl, 30.0)
            self.saver_to_crawl.start()
            while True:
                self.frontier.distribute()
        except:
            self.dash.print_frontier_stat("saving before exit")
            self.frontier.save_to_crawl()


class ControllerRevisit:
    """The Controller of the hole Revisiting process"""

    def __init__(self, num_threads):
        self.num_workers = num_threads
        self.workers = []
        self.frontier = FrontierRevisit(num_threads)
        self.db = Storage()
        # Create the workers
        for i in range(num_threads):
            self.workers.append(RevisiterThread(i, 'CrawlerThread' + str(i), self.frontier))
        print("Workers created")

    def run(self):
        """The main Program"""
        try:
            self.frontier.distribute()
            for i in range(self.num_workers):
                self.workers[i].start()
            print("All Workers started")
            for i in range(self.num_workers):
                self.workers[i].join()
            print("All Workers Finished")
        except KeyboardInterrupt:
            print("Program Interrupted")

File: kokoCrawler/src/Controller.java
public class Controller implements IShutdownThreadParent
{

    public int num_workers;
    public Worker workers[];
    public WorkerSaver saver;
    public Frontier frontier;
    public DB db;
    public Dashboard dash;
    private volatile boolean keepOn = true;

    private ShutdownThread fShutdownThread;


    public Controller(int num_threads, String[] seeds, int mode, int port_no, String to_crawl_db_name)
    {
        num_workers = num_threads;
        workers = new Worker[num_threads];
        dash = new Dashboard();
        db = new DB(port_no);
        frontier = new Frontier(num_threads, db, dash);

        for (int i = 0; i < num_workers; i++)
        {
            workers[i] = new Worker(i, "Thread-" + String.valueOf(i), frontier, db, dash);
        }
        saver = new WorkerSaver(-1, "Thread-Saver", frontier, dash, port_no);
        System.out.println("Workers Created");

        if (mode == 0) // Mode init
        {
            ObjExtractedLink links[] = setup_seeds(seeds);
            frontier.distribute_seeds(links);
            System.out.println("Seeds distributed");
        }
        else if (mode == 1) //Mode cont
        {
            frontier.load_to_crawl(to_crawl_db_name, null);
        }
        //For interrupt
        fShutdownThread = new ShutdownThread(this);
        Runtime.getRuntime().addShutdownHook(fShutdownThread);
    }

    @Override
    public void shutdown()
    {
        // code to cleanly shutdown my Controller.
        try
        {
            for (int i = 0; i < num_workers; i++)
                workers[i].join();
            saver.join();
            System.out.println("All Workers Exit");
        }
        catch (Exception e)
        {
            e.printStackTrace();
        }
        finally
        {
            frontier.save_to_crawl();
            System.out.println("Saved EL7");
        }
    }

    public void run()
    {
        try
        {
            for (int i = 0; i < num_workers; i++)
                workers[i].start();
            saver.start();

            frontier.distribute();

            for (int i = 0; i < num_workers; i++)
                workers[i].join();
            saver.join();
            System.out.println("All Workers Exit");
        }
        catch (Exception e)
        {
            e.printStackTrace();
        }
        finally
        {
            frontier.save_to_crawl();
            System.out.println("Saved EL7");
        }
    }

    public ObjExtractedLink[] setup_seeds(String[] seeds)
    {
        System.out.println("Setting the Seeds");
        ObjLink links[] = new ObjLink[seeds.length];
        int sz = seeds.length;
        for (int i = 0; i < sz; i++)
        {
            links[i] = new ObjLink(seeds[i], Fetcher.extractDNS(seeds[i]));
        }
        ObjExtractedLink ret[] = new ObjExtractedLink[links.length];
        for (int i = 0; i < sz; i++)
        {
            ret[i] = new ObjExtractedLink(links[i], 100000, 100000, 1);
        }
        System.out.println("Seeds are ready");
        System.out.println("Number of seeds = " + String.valueOf(ret.length));
        return ret;
    }
}
